From 0414f7d5bd9383c58fcea0ca7afe3cc3270d3a39 Mon Sep 17 00:00:00 2001
From: Breno Brito <brenorb@gmail.com>
Date: Tue, 6 Oct 2026 00:30:46 -0300
Subject: [PATCH 02/20] refactor(coq): preserve range-separated frames through
 wide and carry writes

Derive all accessed-word separation from initial capacity. Preserve the complete existing 16/32/64-bit and carry execution, output and framing conclusions.

Scoped dependency rebuild and independent kernel check pass. Candidate remains unpublished pending all downstream adaptations and full acceptance.

diff --git a/Coq/C/jet_carry_byte_layout.v b/Coq/C/jet_carry_byte_layout.v
index 7052ecbe60c5b9be921d2c1e4fb0ebea0792d8fb..70428c43334ed9044f8c0b7292eea80ec4892203 100644
--- a/Coq/C/jet_carry_byte_layout.v
+++ b/Coq/C/jet_carry_byte_layout.v
@@ -40,7 +40,8 @@ Theorem eval_carry_byte_layout m bf base bw edge cursor (carry : bool) x :
     (forall b ofs kind p, Mem.perm m b ofs kind p -> Mem.perm mf b ofs kind p) /\
     (forall b, Mem.valid_block m b -> Mem.valid_block mf b).
 Proof.
-  intros HFrame. pose proof HFrame as [HB [HF [HE [HC [HM [HD Hrest]]]]]].
+  intros HFrame. pose proof HFrame as [HB [HF [HE [HC [HM Hrest]]]]].
+  pose proof (write_frame_at_head_separate m bf base bw edge cursor 9 ltac:(lia) HFrame) as HD.
   pose proof (write_layout_index cursor ltac:(lia)) as [_ [HK _]].
   assert (HFrame1 : write_frame_at m bf base bw edge cursor 1).
   { eapply write_frame_at_shorter; [|exact HFrame]; lia. }
@@ -56,7 +57,11 @@ Proof.
   { destruct (Z.eq_dec (write_word_shift cursor) 1) as [Hboundary|Hinside].
     - destruct (write_layout_previous_boundary edge cursor Hboundary) as [Hshift Haddr].
       exists bithigh. split.
-      + rewrite HMem8; [exact HBitLoad|left; congruence|right; right]. rewrite Haddr; lia.
+      + rewrite HMem8; [exact HBitLoad|
+          change (bw <> bf \/ write_word_address edge cursor + 8 <= base + 8 \/
+            base + 16 <= write_word_address edge cursor);
+          destruct HD as [HD|[HD|HD]]; [left; congruence|right; left; lia|right; right; exact HD]
+        |right; right]. rewrite Haddr; lia.
       + intros i Hi Hout; reflexivity.
     - destruct (write_layout_previous_inside edge cursor ltac:(lia)) as [Hshift Haddr].
       unfold write_prefix_at in HPrefix8. rewrite Hshift, Haddr in HPrefix8.
diff --git a/Coq/C/jet_carry_wide_layout.v b/Coq/C/jet_carry_wide_layout.v
index eb9492574640b98cd3c4868db377f5f225b3f8d4..da511259980d55f9e6175994bafb76898489dd14 100644
--- a/Coq/C/jet_carry_wide_layout.v
+++ b/Coq/C/jet_carry_wide_layout.v
@@ -48,7 +48,8 @@ Theorem eval_carry_wide_layout (s : wide_size) m bf base bw edge cursor
 Proof.
   intros HFrame.
   pose proof (wide_bits_bounds s) as HN.
-  pose proof HFrame as [HB [HF [HE [HC [HM [HD [PD HW]]]]]]].
+  pose proof HFrame as [HB [HF [HE [HC [HM [PD HW]]]]]].
+  pose proof (write_frame_at_head_separate m bf base bw edge cursor (1 + wide_bits s) ltac:(lia) HFrame) as HD.
   assert (HFrameNext : write_frame_at m bf base bw edge cursor (wide_bits s + 1)).
   { replace (wide_bits s + 1) with (1 + wide_bits s) by lia. exact HFrame. }
   pose proof (write_layout_index cursor ltac:(lia)) as [_ [HK _]].
@@ -74,7 +75,11 @@ Proof.
   { destruct (Z.eq_dec (write_word_shift cursor) 1) as [Hboundary|Hinside].
     - destruct (write_layout_previous_boundary edge cursor Hboundary) as [Hshift Haddr].
       exists bithigh. split.
-      + rewrite HMemWide; [exact HBitLoad|left; congruence|right; right].
+      + rewrite HMemWide; [exact HBitLoad|
+          change (bw <> bf \/ write_word_address edge cursor + 8 <= base + 8 \/
+            base + 16 <= write_word_address edge cursor);
+          destruct HD as [HD|[HD|HD]]; [left; congruence|right; left; lia|right; right; exact HD]
+        |right; right].
         rewrite Haddr. lia.
       + intros i Hi Hout. reflexivity.
     - destruct (write_layout_previous_inside edge cursor ltac:(lia)) as [Hshift Haddr].
diff --git a/Coq/C/jet_output_layout_step.v b/Coq/C/jet_output_layout_step.v
index 0266605cee7af3d923365d615973284fd69be8dd..8b6636660f05daeabf586b4c34cefabbcff8286a 100644
--- a/Coq/C/jet_output_layout_step.v
+++ b/Coq/C/jet_output_layout_step.v
@@ -9,9 +9,9 @@ Lemma write_frame_at_shorter m bf base bw edge cursor count shorter :
   0 <= shorter <= count ->
   write_frame_at m bf base bw edge cursor count -> write_frame_at m bf base bw edge cursor shorter.
 Proof.
-  intros HS [HB [HF [HE [HC [HM [HD [PD HW]]]]]]].
+  intros HS [HB [HF [HE [HC [HM [PD HW]]]]]].
   split; [exact HB|]. split; [exact HF|]. split; [exact HE|]. split; [lia|].
-  split; [exact HM|]. split; [exact HD|]. split; [exact PD|]. intros i Hi; apply HW; lia.
+  split; [exact HM|]. split; [exact PD|]. intros i Hi; apply HW; lia.
 Qed.
 
 Lemma write_frame_at_after_bit m mf bf base bw edge cursor count w :
@@ -23,19 +23,22 @@ Lemma write_frame_at_after_bit m mf bf base bw edge cursor count w :
   (forall b ofs kind p, Mem.perm m b ofs kind p -> Mem.perm mf b ofs kind p) ->
   write_frame_at mf bf base bw edge (cursor - 1) count.
 Proof.
-  intros Hcount [HB [HF [HE [HC [HM [HD [PD HW]]]]]]] HF' Hhead Houtside HP.
+  intros Hcount [HB [HF [HE [HC [HM [PD HW]]]]]] HF' Hhead Houtside HP.
   assert (HV : forall chunk b ofs p, Mem.valid_access m chunk b ofs p -> Mem.valid_access mf chunk b ofs p).
   { intros chunk b ofs p [HR HA]. split; [|exact HA]. intros addr HA'. apply HP. apply HR; exact HA'. }
   split; [exact HB|]. split; [exact HF'|]. split; [exact HE|]. split; [lia|].
-  split; [lia|]. split; [exact HD|]. split; [auto|].
+  split; [lia|]. split; [auto|].
   intros i Hi. cbn zeta.
   assert (Haddr : write_cell_address edge (cursor - 1) i = write_cell_address edge cursor (i + 1)).
   { unfold write_cell_address. replace (cursor - 1 - 1 - i) with (cursor - 1 - (i + 1)) by lia. reflexivity. }
-  rewrite Haddr. destruct (HW (i + 1) ltac:(lia)) as [HA0 [HA [PW [old HL]]]].
-  split; [exact HA0|]. split; [exact HA|]. split; [auto|].
+  rewrite Haddr. destruct (HW (i + 1) ltac:(lia)) as [HA0 [HA [HDataSep [PW [old HL]]]]].
+  split; [exact HA0|]. split; [exact HA|]. split; [exact HDataSep|]. split; [auto|].
+  assert (Hcursorsep : bw <> bf \/ write_cell_address edge cursor (i + 1) + 8 <= base + 8 \/
+      base + 16 <= write_cell_address edge cursor (i + 1)).
+  { destruct HDataSep as [HD|[HD|HD]]; [left; congruence|right; left; lia|right; right; exact HD]. }
   destruct (Z.eq_dec (write_cell_address edge cursor (i + 1)) (write_word_address edge cursor)) as [Heq | Hneq].
   - rewrite Heq. exists w; exact Hhead.
-  - exists old. rewrite Houtside; [exact HL|left; congruence|].
+  - exists old. rewrite Houtside; [exact HL|exact Hcursorsep|].
     right. unfold write_cell_address, write_word_address in Hneq |- *.
     change (edge + 8 * ((cursor - 1 - (i + 1)) / 64) + 8 <= edge + 8 * ((cursor - 1) / 64) \/
       edge + 8 * ((cursor - 1) / 64) + 8 <= edge + 8 * ((cursor - 1 - (i + 1)) / 64)).
diff --git a/Coq/C/jet_output_slice.v b/Coq/C/jet_output_slice.v
index c3149ae0bdbce6b53825b973d155974989fb4332..0f69caf60f2de14458e0bdc2959ed5c873781bb1 100644
--- a/Coq/C/jet_output_slice.v
+++ b/Coq/C/jet_output_slice.v
@@ -47,10 +47,23 @@ Lemma write_frame_at_slice_crossing n m bf base bw edge cursor :
   Mem.valid_access m Mint64 bw (write_word_address edge cursor - 8) Writable /\
   exists w, Mem.load Mint64 m bw (write_word_address edge cursor - 8) = Some (Vlong w).
 Proof.
-  intros Hcross [_ [_ [_ [HC [HM [_ [_ HW]]]]]]].
+  intros Hcross (HB & HF & HE & HC & HM & PF & HW).
   assert (HK : 1 <= write_word_shift cursor <= 64).
   { unfold write_word_shift. pose proof (Z.mod_pos_bound (cursor - 1) 64 ltac:(lia)); lia. }
   specialize (HW (write_word_shift cursor) ltac:(lia)). cbn zeta in HW.
   rewrite slice_crossing_address in HW by reflexivity.
-  destruct HW as [HL [_ HW]]. split; [lia|exact HW].
+  destruct HW as (HA0 & HA & HD & PW & HL). split; [lia|]. split; assumption.
+Qed.
+
+Lemma write_frame_at_slice_crossing_separate n m bf base bw edge cursor :
+  write_word_shift cursor < n -> write_frame_at m bf base bw edge cursor n ->
+  bf <> bw \/ write_word_address edge cursor - 8 + 8 <= base \/
+    base + 16 <= write_word_address edge cursor - 8.
+Proof.
+  intros Hcross (HB & HF & HE & HC & HM & PF & HW).
+  assert (HK : 1 <= write_word_shift cursor <= 64).
+  { unfold write_word_shift. pose proof (Z.mod_pos_bound (cursor - 1) 64 ltac:(lia)); lia. }
+  specialize (HW (write_word_shift cursor) ltac:(lia)). cbn zeta in HW.
+  rewrite slice_crossing_address in HW by reflexivity.
+  exact (proj1 (proj2 (proj2 HW))).
 Qed.
diff --git a/Coq/C/jet_write_wide_layout.v b/Coq/C/jet_write_wide_layout.v
index 16a838ed3661559e42bca1fc8cf10bba61bcc682..f545e24f414768d35ddc4b1b9b1d9771138e3437 100644
--- a/Coq/C/jet_write_wide_layout.v
+++ b/Coq/C/jet_write_wide_layout.v
@@ -117,7 +117,9 @@ Lemma eval_write_wide_non_crossing_raw s m mw mf bf base bw edge cursor old x :
   Mem.load Mint64 m bw (write_word_address edge cursor) = Some (Vlong old) ->
   Mem.store Mint64 m bw (write_word_address edge cursor)
     (Vlong (put_slice (wide_bits s) (write_word_shift cursor) old x)) = Some mw ->
-  Mem.store Mint64 mw bf (base + 8) (Vlong (Int64.repr (cursor - wide_bits s))) = Some mf -> bf <> bw ->
+  Mem.store Mint64 mw bf (base + 8) (Vlong (Int64.repr (cursor - wide_bits s))) = Some mf ->
+  (bf <> bw \/ write_word_address edge cursor + 8 <= base \/
+    base + 16 <= write_word_address edge cursor) ->
   ClightBigstep.Clight2.eval_funcall ge0 m (Internal (wide_writer s))
     [Vptr bf (Ptrofs.repr base); Vlong x] E0 mf Vundef.
 Proof.
@@ -138,7 +140,10 @@ Proof.
       (Vlong (put_slice (wide_bits s) (write_word_shift cursor) old x)) = Some mw)
     by (rewrite Haddr; exact SW).
   assert (HOw : Mem.load Mint64 mw bf (base + 8) = Some (Vlong (Int64.repr cursor))).
-  { erewrite Mem.load_store_other; [exact HO|exact SW|auto]. }
+  { erewrite Mem.load_store_other; [exact HO|exact SW|].
+    change (bf <> bw \/ base + 8 + 8 <= write_word_address edge cursor \/
+      write_word_address edge cursor + 8 <= base + 8).
+    destruct HD as [HD|[HD|HD]]; [left; exact HD|right; right; lia|right; left; lia]. }
   assert (Hcross : Int64.ltu (Int64.repr (write_word_shift cursor)) (Int64.repr (wide_bits s)) = false).
   { unfold Int64.ltu. rewrite !cursor_unsigned by lia. rewrite zlt_false by lia; reflexivity. }
   pose proof (cursor_sub (write_word_shift cursor) (wide_bits s) ltac:(lia) ltac:(lia)) as Hshift_sub.
diff --git a/Coq/C/jet_write_wide_layout_total.v b/Coq/C/jet_write_wide_layout_total.v
index 63d136da486fea89721025ad899d8c2b4bb75f55..0d835db0091a66227efeeeead2d26ac7bc8f1184 100644
--- a/Coq/C/jet_write_wide_layout_total.v
+++ b/Coq/C/jet_write_wide_layout_total.v
@@ -20,9 +20,10 @@ Theorem eval_write_wide_layout s m bf base bw edge cursor x :
     (forall b, Mem.valid_block m b -> Mem.valid_block mf b).
 Proof.
   intros HFrame. pose proof (wide_bits_bounds s) as HNbound.
-  pose proof HFrame as [HB [[HF HO] [HE [HC [HM [HD [PD HW]]]]]]].
+  pose proof HFrame as [HB [[HF HO] [HE [HC [HM [PD HW]]]]]].
   destruct (write_frame_at_head m bf base bw edge cursor (wide_bits s) ltac:(lia) HFrame)
     as [HA0 [HA [PH [oldhigh HH]]]].
+  pose proof (write_frame_at_head_separate m bf base bw edge cursor (wide_bits s) ltac:(lia) HFrame) as HD.
   pose proof (write_layout_index cursor ltac:(lia)) as [_ [HK _]].
   unfold slice_output_at, slice_write_low.
   destruct (Z_le_dec (wide_bits s) (write_word_shift cursor)) as [HN | HX].
@@ -34,7 +35,7 @@ Proof.
     destruct (Mem.valid_access_store mw Mint64 bf (base + 8) (Vlong (Int64.repr (cursor - wide_bits s))) PDw)
       as [mf SF].
     assert (Houtput : Mem.load Mint64 mf bw (write_word_address edge cursor) = Some (Vlong output)).
-    { erewrite Mem.load_store_other; [|exact SF|auto]. exact (Mem.load_store_same _ _ _ _ _ _ SW). }
+    { erewrite frame_store_preserves_word_load; [|eassumption|exact SF]. exact (Mem.load_store_same _ _ _ _ _ _ SW). }
     exists mf. split.
     + eapply eval_write_wide_non_crossing_raw with (edge := edge) (old := oldhigh);
         eauto; try lia. split; assumption.
@@ -47,7 +48,7 @@ Proof.
         -- split.
            ++ split.
               ** erewrite Mem.load_store_other; [|exact SF|right; left; change (base + 8 <= base + 8); lia].
-                 erewrite Mem.load_store_other; [exact HF|exact SW|auto].
+                 erewrite word_store_preserves_frame_load; [exact HF|exact HD|lia|change (base + 8 <= base + 16); lia|exact SW].
               ** exact (Mem.load_store_same _ _ _ _ _ _ SF).
            ++ split.
               ** intros chunk b ofs Hbf Hbw.
@@ -58,6 +59,7 @@ Proof.
                  --- intros b HV. eauto using Mem.store_valid_block_1.
   - destruct (write_frame_at_slice_crossing (wide_bits s) m bf base bw edge cursor ltac:(lia) HFrame)
       as [HAL [PL [oldlow HL]]].
+    pose proof (write_frame_at_slice_crossing_separate (wide_bits s) m bf base bw edge cursor ltac:(lia) HFrame) as HDL.
     set (k := write_word_shift cursor).
     destruct (Mem.valid_access_store m Mint64 bw (write_word_address edge cursor)
         (Vlong (slice_high (wide_bits s) k oldhigh x)) PH) as [mh SH].
@@ -74,20 +76,20 @@ Proof.
     destruct (Mem.valid_access_store mw Mint64 bf (base + 8)
         (Vlong (Int64.repr (cursor - wide_bits s))) PDw) as [mf SF].
     assert (Hhigh : Mem.load Mint64 mf bw (write_word_address edge cursor) = Some (Vlong (slice_high (wide_bits s) k oldhigh x))).
-    { erewrite Mem.load_store_other; [|exact SF|auto].
+    { erewrite frame_store_preserves_word_load; [|eassumption|exact SF].
       erewrite Mem.load_store_other; [|exact SW|right; right; cbn; lia].
-      erewrite Mem.load_store_other; [|exact SO|auto].
+      erewrite frame_store_preserves_word_load; [|eassumption|exact SO].
       exact (Mem.load_store_same _ _ _ _ _ _ SH). }
     assert (Hlow : Mem.load Mint64 mf bw (write_word_address edge cursor - 8) = Some (Vlong (slice_low (wide_bits s) k x))).
-    { erewrite Mem.load_store_other; [|exact SF|auto]. exact (Mem.load_store_same _ _ _ _ _ _ SW). }
+    { erewrite frame_store_preserves_word_load; [|eassumption|exact SF]. exact (Mem.load_store_same _ _ _ _ _ _ SW). }
     exists mf. split.
     + eapply eval_write_wide_crossing_raw with (edge := edge) (k := k)
         (oldhigh := oldhigh) (oldlow := oldlow); eauto; try (unfold k; lia).
       * split; assumption.
-      * erewrite Mem.load_store_other; [exact HO|exact SH|auto].
-      * erewrite Mem.load_store_other; [|exact SO|auto].
+      * erewrite word_store_preserves_frame_load; [exact HO|exact HD|lia|change (base + 8 + 8 <= base + 16); lia|exact SH].
+      * erewrite frame_store_preserves_word_load; [|eassumption|exact SO].
         erewrite Mem.load_store_other; [exact HL|exact SH|right; left; cbn; lia].
-      * erewrite Mem.load_store_other; [|exact SW|auto]. exact (Mem.load_store_same _ _ _ _ _ _ SO).
+      * erewrite word_store_preserves_frame_load; [|exact HDL|lia|change (base + 8 + 8 <= base + 16); lia|exact SW]. exact (Mem.load_store_same _ _ _ _ _ _ SO).
     + split.
       * exists (slice_high (wide_bits s) k oldhigh x), (slice_low (wide_bits s) k x).
         split; [exact Hhigh|]. split; [exact Hlow|]. apply crossing_slice_projection; unfold k; lia.
@@ -97,9 +99,9 @@ Proof.
         -- split.
            ++ split.
               ** erewrite Mem.load_store_other; [|exact SF|right; left; change (base + 8 <= base + 8); lia].
-                 erewrite Mem.load_store_other; [|exact SW|auto].
+                 erewrite word_store_preserves_frame_load; [|exact HDL|lia|change (base + 8 <= base + 16); lia|exact SW].
                  erewrite Mem.load_store_other; [|exact SO|right; left; change (base + 8 <= base + 8); lia].
-                 erewrite Mem.load_store_other; [exact HF|exact SH|auto].
+                 erewrite word_store_preserves_frame_load; [exact HF|exact HD|lia|change (base + 8 <= base + 16); lia|exact SH].
               ** exact (Mem.load_store_same _ _ _ _ _ _ SF).
            ++ split.
               ** intros chunk b ofs Hbf Hbw.
